File: SwrveSDK/SDK/Messaging/SwrveInAppCampaign.m
#import "Swrve.h"
#import "SwrveInAppCampaign.h"
#import "SwrveButton.h"
#import "SwrveImage.h"

#if __has_include(<SwrveSDKCommon/SwrveAssetsManager.h>)

#import <SwrveSDKCommon/SwrveAssetsManager.h>
#import <SwrveSDKCommon/SwrveUtils.h>
#import <SwrveSDKCommon/TextTemplating.h>

#else
#import "SwrveAssetsManager.h"
#import "SwrveUtils.h"
#import "TextTemplating.h"
#endif

#import "SwrveCampaign+Private.h"
#import "SwrveMessageController+Private.h"
#import "SwrveMessagePage.h"

@implementation SwrveInAppCampaign

@synthesize message;

- (id)initAtTime:(NSDate *)time fromDictionary:(NSDictionary *)json withAssetsQueue:(NSMutableSet *)assetsQueue forController:(SwrveMessageController *)controller withPersonalization:(NSDictionary *)personalization {
    id instance = [super initAtTime:time fromDictionary:json];
    NSDictionary *messageDict = [json objectForKey:@"message"];
    self.message = [[SwrveMessage alloc] initWithDictionary:messageDict forCampaign:self forController:controller];
    [self addAssetsToQueue:assetsQueue withPersonalization:personalization];
    if ([json objectForKey:@"subject"] != [NSNull null] && [json objectForKey:@"subject"] != nil) {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
        self.subject = [json objectForKey:@"subject"];
#pragma clang diagnostic pop
    }
    
    self.name = self.message.name;

    self.campaignType = SWRVE_CAMPAIGN_IAM;

    return instance;
}

- (void)addAssetToQueue:(NSMutableSet *)assetsQueue withUrl:(NSString *)url withPersonalization:(NSDictionary *)personalization {
    if (url == nil) return;
    NSError *error;
    NSString *resolvedUrl = [TextTemplating templatedTextFromString:url withProperties:personalization andError:&error];
    if (error != nil || resolvedUrl == nil) {
        [SwrveLogger debug:@"Could not resolve personalization: %@", url];
    } else {
        NSData *data = [resolvedUrl dataUsingEncoding:NSUTF8StringEncoding allowLossyConversion:YES];
        NSString *sha1Url = [SwrveUtils sha1:data];
        NSMutableDictionary *assetQueueItem = [SwrveAssetsManager assetQItemWith:sha1Url andDigest:resolvedUrl andIsExternal:YES andIsImage:YES];
        [assetsQueue addObject:assetQueueItem];
    }
}

- (void)addAssetsToQueue:(NSMutableSet *)assetsQueue withPersonalization:(NSDictionary *)personalization {
    NSString *imageUrl = self.message.messageCenterDetails.imageUrl;
    if (imageUrl != nil) {
        [self addAssetToQueue:assetsQueue withUrl:imageUrl withPersonalization:personalization];
    }
    
    NSString *imageSha = self.message.messageCenterDetails.imageSha;
    if (imageSha != nil) {
        NSMutableDictionary *assetQueueItem = [SwrveAssetsManager assetQItemWith:imageSha andDigest:imageSha andIsExternal:NO andIsImage:YES];
        [assetsQueue addObject:assetQueueItem];
    }
    
    for (SwrveMessageFormat *format in message.formats) {
        NSDictionary *pages = [format pages];
        for (id key in pages) {
            SwrveMessagePage *page = [pages objectForKey:key];

            for (SwrveButton *button in page.buttons) {
                [self addAssetToQueue:assetsQueue withUrl:button.dynamicImageUrl withPersonalization:personalization];

                if (button.image) {
                    NSMutableDictionary *assetQueueItem = [SwrveAssetsManager assetQItemWith:button.image andDigest:button.image andIsExternal:NO andIsImage:YES];
                    [assetsQueue addObject:assetQueueItem];
                }
            }

            for (SwrveImage *image in page.images) {
                [self addAssetToQueue:assetsQueue withUrl:image.dynamicImageUrl withPersonalization:personalization];
                
                if (image.file) {
                    NSMutableDictionary *assetQueueItem = [SwrveAssetsManager assetQItemWith:image.file andDigest:image.file andIsExternal:NO andIsImage:YES];
                    [assetsQueue addObject:assetQueueItem];
                }

                if (image.multilineText) {
                    [self addFontToQ:assetsQueue withAsset:image.multilineText];
                }
            }
        }
    }
}

- (void)addFontToQ:(NSMutableSet *)assetQueue withAsset:(NSDictionary *)style {
    NSString *name = [style objectForKey:@"font_file"];
    NSString *digest = [style objectForKey:@"font_digest"];
    if (name && digest && ![SwrveBaseConversation isSystemFont:style]) {
        NSMutableDictionary *assetQueueItem = [SwrveAssetsManager assetQItemWith:name andDigest:digest andIsExternal:NO andIsImage:NO];
        [assetQueue addObject:assetQueueItem];
    }
}

- (void)messageWasShownToUser:(SwrveMessage *)messageShown {
    [self wasShownToUserAt:[NSDate date]];
}

- (void)messageDismissed:(NSDate *)timeDismissed {
    [self setMessageMinDelayThrottle:timeDismissed];
}

/**
 * Quick check to see if this campaign might have messages matching this event trigger
 * This is used to decide if the campaign is a valid candidate for automatically showing at session start
 */
- (BOOL)hasMessageForEvent:(NSString *)event {
    return [self hasMessageForEvent:event withPayload:nil];
}

- (BOOL)hasMessageForEvent:(NSString *)event withPayload:(NSDictionary *)payload {
    return [self canTriggerWithEvent:event andPayload:payload];
}

- (SwrveMessage *)messageForEvent:(NSString *)event
                       withAssets:(NSSet *)assets
              withPersonalization:(NSDictionary *)personalization
                           atTime:(NSDate *)time {
    return [self messageForEvent:event withPayload:nil withAssets:assets withPersonalization:personalization atTime:time withReasons:nil];
}

- (SwrveMessage *)messageForEvent:(NSString *)event
                      withPayload:(NSDictionary *)payload
                       withAssets:(NSSet *)assets
              withPersonalization:(NSDictionary *)personalization
                           atTime:(NSDate *)time
                      withReasons:(NSMutableDictionary *)campaignReasons {

    if (![self hasMessageForEvent:event withPayload:payload]) {
        [SwrveLogger debug:@"There is no trigger in %ld that matches %@", (long) self.ID, event];
        [self logAndAddReason:[NSString stringWithFormat:@"There is no trigger in %ld that matches %@ with conditions %@", (long) self.ID, event, payload] withReasons:campaignReasons];
        return nil;
    }

    if (self.message == nil) {
        [self logAndAddReason:[NSString stringWithFormat:@"No messages in campaign %ld", (long) self.ID] withReasons:campaignReasons];
        return nil;
    }

    if (![self checkCampaignRulesForEvent:event atTime:time withReasons:campaignReasons]) {
        return nil;
    }

    if ([message assetsReady:assets withPersonalization:personalization]) {
        [SwrveLogger debug:@"%@ matches a trigger in %ld", event, (long) self.ID];
        return message;
    }

    [self logAndAddReason:[NSString stringWithFormat:@"Campaign %ld hasn't finished downloading", (long) self.ID] withReasons:campaignReasons];
    return nil;
}

#if TARGET_OS_IOS /** exclude tvOS **/

- (BOOL)supportsOrientation:(UIInterfaceOrientation)orientation {
    if (orientation == UIInterfaceOrientationUnknown) {
        return YES;
    }
    if ([self.message supportsOrientation:orientation]) {
        return YES;
    }
    return NO;
}

#endif

- (BOOL)assetsReady:(NSSet *)assets withPersonalization:(NSDictionary *)personalization {
    if (![message assetsReady:assets withPersonalization:personalization]) {
        return NO;
    }
    return YES;
}

@end

File: SwrveSDKSwift/SwrveInAppCampaign.swift
import Foundation
import UIKit

#if canImport(SwrveSDK)
    import SwrveSDK
#endif

#if canImport(SwrveSDKCommon)
    import SwrveSDKCommon
#endif

@objc public class SwrveInAppCampaign: SwrveCampaign {

    /// Message attached to this campaign.
    @objc public var message: SwrveMessage?

    @objc public init(
        atTime time: Date, fromDictionary json: [String: Any], withAssetsQueue assetsQueue: NSMutableSet,
        forController controller: SwrveMessageController, withPersonalization personalization: [String: Any]
    ) {
        super.init(at: time, from: json, campaignType: SWRVE_CAMPAIGN_IAM)
        if let messageDict = json["message"] as? [String: Any] {
            self.message = SwrveMessage(dictionary: messageDict, campaign: self, controller: controller)
            self.addAssetsToQueue(assetsQueue, withPersonalization: personalization)
            self.name = self.message?.name ?? ""
        }
    }

    func addAssetToQueue(assetsQueue: NSMutableSet, withUrl url: String?, withPersonalization personalization: [String: Any]) {
        guard let url = url else { return }
        do {

            let resolvedUrl = try TextTemplating.templatedText(from: url, withProperties: personalization)
            let data = resolvedUrl.data(using: .utf8, allowLossyConversion: true)
            let sha1Url = SwrveUtils.sha1(data)
            if let assetQueueItem = SwrveAssetsManager.assetQItem(with: sha1Url, andDigest: resolvedUrl, andIsExternal: true, andIsImage: true) {
                assetsQueue.add(assetQueueItem)
            }
        } catch {
            SwrveLogger.logDebug("Could not resolve personalization: \(url)")
        }
    }

    @objc public func addAssetsToQueue(_ assetsQueue: NSMutableSet, withPersonalization personalization: [String: Any]) {
        if let imageURL = message?.messageCenterDetails?.imageUrl {
            addAssetToQueue(assetsQueue: assetsQueue, withUrl: imageURL, withPersonalization: personalization)
        }

        if let imageSha = message?.messageCenterDetails?.imageSha {
            if let assetQueueItem = SwrveAssetsManager.assetQItem(with: imageSha, andDigest: imageSha, andIsExternal: false, andIsImage: true) {
                assetsQueue.add(assetQueueItem)
            }
        }

        message?.formats.forEach { format in
            format.pages.forEach { (_, page) in

                if let page = page as? SwrveMessagePage {
                    page.buttons.forEach { button in

                        if let button = button as? SwrveButton {
                            addAssetToQueue(assetsQueue: assetsQueue, withUrl: button.dynamicImageUrl, withPersonalization: personalization)

                            if let image = button.image,
                                let assetQueueItem = SwrveAssetsManager.assetQItem(
                                    with: image, andDigest: image, andIsExternal: false, andIsImage: true)
                            {
                                assetsQueue.add(assetQueueItem)
                            }

                            if let theme = button.theme {
                                addButtonThemeAssets(theme: theme, assetsQueue: assetsQueue)
                            }
                        }
                    }

                    page.images.forEach { image in

                        if let image = image as? SwrveImage {
                            addAssetToQueue(assetsQueue: assetsQueue, withUrl: image.dynamicImageUrl, withPersonalization: personalization)

                            if let file = image.file,
                                let assetQueueItem = SwrveAssetsManager.assetQItem(
                                    with: file, andDigest: file, andIsExternal: false, andIsImage: true)
                            {
                                assetsQueue.add(assetQueueItem)
                            }

                            if let multilineText = image.multilineText as? [String: Any] {
                                addFontToQueue(assetQueue: assetsQueue, withAsset: multilineText)
                            }
                        }
                    }
                }
            }
        }
    }

    func addButtonThemeAssets(theme: SwrveButtonTheme, assetsQueue: NSMutableSet) {
        if let bgImage = theme.bgImage,
            let assetQueueItem = SwrveAssetsManager.assetQItem(with: bgImage, andDigest: bgImage, andIsExternal: false, andIsImage: true)
        {
            assetsQueue.add(assetQueueItem)
        }

        let fontFile = theme.fontFile

        if !SwrveSDKUtils.isSystemFont(fontFile),
            let assetQueueItem = SwrveAssetsManager.assetQItem(with: fontFile, andDigest: theme.fontDigest, andIsExternal: false, andIsImage: false)
        {
            assetsQueue.add(assetQueueItem)
        }

        if let pressedState = theme.pressedState,
            let bgImage = pressedState.bgImage,
            let assetQueueItem = SwrveAssetsManager.assetQItem(with: bgImage, andDigest: bgImage, andIsExternal: false, andIsImage: true)
        {
            assetsQueue.add(assetQueueItem)
        }

        if let focusedState = theme.focusedState,
            let bgImage = focusedState.bgImage,
            let assetQueueItem = SwrveAssetsManager.assetQItem(with: bgImage, andDigest: bgImage, andIsExternal: false, andIsImage: true)
        {
            assetsQueue.add(assetQueueItem)
        }
    }

    func addFontToQueue(assetQueue: NSMutableSet, withAsset style: [String: Any]) {
        if let name = style["font_file"] as? String,
            let digest = style["font_digest"] as? String,
            let assetQueueItem = SwrveAssetsManager.assetQItem(with: name, andDigest: digest, andIsExternal: false, andIsImage: false)
        {
            assetQueue.add(assetQueueItem)
        }
    }

    func messageWasShownToUser(_ messageShown: SwrveMessage) {
        super.wasShownToUser(at: Date())
    }

    @objc public func messageDismissed(_ timeDismissed: Date) {
        super.setMessageMinDelayThrottle(at: timeDismissed)
    }

    /**
     Quick check to see if this campaign might have messages matching this event trigger.
     This is used to decide if the campaign is a valid candidate for automatically showing at session start.

     - Parameter event: Trigger event.
     - Returns: `true` if the campaign contains a message for the given trigger.
     */
    func hasMessage(forEvent event: String) -> Bool {
        hasMessage(forEvent: event, withPayload: nil)
    }

    @objc public func hasMessage(forEvent event: String, withPayload payload: [AnyHashable: Any]?) -> Bool {
        super.canTrigger(withEvent: event, andPayload: payload)
    }

    func message(
        forEvent event: String, withAssets assets: Set<AnyHashable>, withPersonalization personalization: [String: Any], atTime time: Date
    ) -> SwrveMessage? {
        message(
            forEvent: event, withPayload: nil, withAssets: assets, withPersonalization: personalization, atTime: time,
            withReasons: NSMutableDictionary())
    }

    @objc public func message(
        forEvent event: String, withPayload payload: [AnyHashable: Any]?, withAssets assets: Set<AnyHashable>,
        withPersonalization personalization: [String: Any], atTime time: Date, withReasons campaignReasons: NSMutableDictionary
    ) -> SwrveMessage? {

        guard hasMessage(forEvent: event, withPayload: payload) else {
            SwrveLogger.logDebug("There is no trigger in \(self.ID) that matches \(event)")
            logAndAdd(
                reason:
                    "There is no trigger in \(self.ID) that matches \(event) with conditions \(String(describing: payload))",
                withReasons: campaignReasons
            )
            return nil
        }

        guard let message = self.message else {
            logAndAdd(reason: "No messages in campaign \(self.ID)", withReasons: campaignReasons)
            return nil
        }
        guard checkCampaignRules(forEvent: event, atTime: time, withReasons: campaignReasons) else {
            return nil
        }

        let assetsAsSet = (assets as? Set<String>) ?? Set<String>()
        if message.assetsReady(assetsAsSet, withPersonalization: personalization) {
            SwrveLogger.logDebug("\(event) matches a trigger in \(self.ID)")
            return message
        }

        logAndAdd(reason: "Campaign \(self.ID) hasn't finished downloading", withReasons: campaignReasons)
        return nil
    }

    #if os(iOS)
        @objc public override func supportsOrientation(_ orientation: UIInterfaceOrientation) -> Bool {
            if orientation == .unknown {
                return true
            }

            if message!.supportsOrientation(orientation) {
                return true
            }

            return message?.supportsOrientation(orientation) ?? false
        }
    #endif

    @objc override public func assetsReady(_ assets: Set<String>, withPersonalization personalization: [String: Any]) -> Bool {
        guard let message = message else {
            return true
        }
        if !message.assetsReady(assets, withPersonalization: personalization) {
            return false
        }
        return true
    }
}
